cirq/value/states.py

```diff
@@ -0,0 +1,365 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Modifications are covered under:
+#
+# Copyright 2020 The Cirq developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from dataclasses import dataclass
+from typing import Iterable, TYPE_CHECKING, Dict
+
+import numpy as np
+
+from cirq import protocols
+from cirq._doc import document
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class _Named1QState(metaclass=abc.ABCMeta):
+    """Abstract class representing a one-qubit state of note."""
+
+    def on(self, qubit) -> 'TensorProductState':
+        """Associates one qubit with this named state.
+
+        The returned object is a TensorProductState of length 1.
+        """
+        return TensorProductState({qubit: self})
+
+    def __call__(self, *args, **kwargs):
+        return self.on(*args, **kwargs)
+
+    @abc.abstractmethod
+    def state_vector(self):
+        """Return a state vector representation of the named state."""
+
+    def projector(self):
+        """Return |s><s| as a matrix for the named state."""
+        vec = self.state_vector()[:, np.newaxis]
+        return vec @ vec.conj().T
+
+    def _json_dict_(self):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return protocols.obj_to_dict_helper(self, [])
+
+    def __eq__(self, other):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return self.__class__ == other.__class__
+
+    def __hash__(self):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return hash(self.__class__.__name__)
+
+
+@dataclass(frozen=True)
+class TensorProductState:
+    """A quantum state that is a tensor product of one qubit states.
+
+    For example, the |00> state is `cirq.PLUS_Z(q0) * cirq.PLUS_Z(q1)`.
```

Comment:
I'm a bit worried about the names here. There's an apparent similarly, but large difference, between `+cirq.Z(q0) * +cirq.Z(q1)` and `cirq.PLUS_Z(q0) * cirq.PLUS_Z(q1)`. It's also not the case that cirq.MINUS_Z(q) is equal to -cirq.PLUS_Z(q).

(In somewhat related news I foolishly initially thought that PLUS_Z was referring to the plus state |+> with respect to the Z basis.)

I don't have a better suggestion at the moment, just this worry about these names leading to predictable confusions.

Some ideas:

- Prefix the names with KET_ to more clearly mark them as states, and break the "plus_x equals -minus_x" connotation.
- Switch to a completely different strategy, like a dictionary from qubits to plus or minus cirq.X/Y/Z.
- Have a method like "PlusEigenstate" and do stuff like `PlusEigenstate(cirq.X(q)) * PlusEigenstate(-cirq.Z(q))`
- Give Pauli gates a helper method `.eigen_ket()` and do stuff like `cirq.X(q).eigen_ket(+1) * cirq.Z(q).eigen_ket(-1)`.
- Generalize to StabilizerState and take a list of stabilizers (with the single qubit stabilizers as a special case).
- Use names like `KET_ZERO`, `KET_PLUS`, and `KET_IMAG`.

Reply:
> Switch to a completely different strategy, like a dictionary from qubits to plus or minus cirq.X/Y/Z.

This is pretty close to what we're actually doing, but the dictionary is to `_Named1QState` instead. The thought is that it would be easier to extend this to other states of note (like SIC basis for tomography; magic states, ...)

> Prefix the names with KET_ to more clearly mark them as states, and break the "plus_x equals -minus_x" connotation.
> Use names like KET_ZERO, KET_PLUS, and KET_IMAG.

These two suggestions seem similar and I will do the latter

> Have a method like "PlusEigenstate" and do stuff like PlusEigenstate(cirq.X(q)) * PlusEigenstate(-cirq.Z(q))
> Give Pauli gates a helper method .eigen_ket() and do stuff like cirq.X(q).eigen_ket(+1) * cirq.Z(q).eigen_ket(-1).

These seem a little complicated and verbose and stem from a stance that the `Pauli` object is the be-all-end-all and all other objects should kneel before them. I think the concept of a single-qubit state with a special name is distinct enough that it can exist as a 1st class object (and you can get back to pauli's via the `stabilized_by` method).